### PR TITLE
Dockerfile proposal for user

### PR DESCRIPTION
**Pull request by JuanRodenas**
Dockerfile change proposal with user over docker-compose.
Change requested in issue #1870.

Please review the changes and indicate what we can fix.
- Since we have to create the folders first.
- I see that when sending to the /wiki folder a file, gollum does not compile it and it does not appear in Overview.
- We can also see to get the logs out of the container and send them to syslog

In this fork we modify:
- We removed the Dockerfile to run as root.
- We have added REDIS to the default config.
- We added HEALTHCHECK to the Dockerfile.

I am waiting for news to continue with the project. I request to be a collaborator. 
Best regards, Juan

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,11 +1,15 @@
-FROM ruby:2.7-alpine AS builder
-
-RUN apk add \
-    build-base \
-    cmake \
-    git \
-    icu-dev \
-    openssl-dev
+FROM ruby:3.0-alpine AS builder
+LABEL author="juan rodenas"
```
Comment (dometto) [inline Dockerfile:2]: please remove from this PR

File: Dockerfile [hunk 2]
```diff
@@ -25,17 +29,33 @@ WORKDIR /app
...
+RUN chmod +x /docker-run.sh
+USER www-data
+VOLUME /wiki
+EXPOSE 4567
+
+HEALTHCHECK --start-period=2s --interval=5m --timeout=3s \
+  CMD curl -f http://localhost:4567 || exit 1
+
 ENTRYPOINT ["/docker-run.sh"]
+CMD ["/bin/sh", "/docker-run.sh"]
```
Comment (dometto) [inline Dockerfile:61]: What is the motivation for this change?

File: Dockerfile [hunk 3]
```diff
@@ -25,17 +29,33 @@ WORKDIR /app
...
+    && groupmod -g $GID www-data \
+    && adduser -u $UID -S www-data -G www-data
+
 COPY docker-run.sh /docker-run.sh
+RUN chmod +x /docker-run.sh
+USER www-data
+VOLUME /wiki
+EXPOSE 4567
+
+HEALTHCHECK --start-period=2s --interval=5m --timeout=3s \
```
Comment (dometto) [inline Dockerfile:57]: I am not strictly against adding a `HEALTCHECK` but would prefer to discuss the need, proper implementation, and possible downsides in a separate issue.

File: Dockerfile [hunk 4]
```diff
@@ -25,17 +29,33 @@ WORKDIR /app
...
+            libc6-compat \
+    && rm -rf /var/cache/apk/* \
+    && groupmod -g $GID www-data \
+    && adduser -u $UID -S www-data -G www-data
+
 COPY docker-run.sh /docker-run.sh
+RUN chmod +x /docker-run.sh
+USER www-data
+VOLUME /wiki
+EXPOSE 4567
```
Comment (dometto) [inline Dockerfile:55]: Why is the explicit `EXPOSE` required? Does this have implications for the [documented way of running the Docker image](https://github.com/gollum/gollum/wiki/Gollum-via-Docker)?

**Review by dometto — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 3 above)
Comment (bartkamphorst) [reply Dockerfile:57]: Agreed.

File: Dockerfile (hunk 4 above)
Comment (JuanRodenas) [reply Dockerfile:55]: This section is implicit for the HEALTHCHECK. To perform the curl.

File: Dockerfile (hunk 1 above)
Comment (JuanRodenas) [reply Dockerfile:2]: Ok, I just put this LABEL for my Github

**Comment by JuanRodenas** [comment]
Thank you very much for the acceptance of the PR, I will remove the LABEL, Redis. Upgrading to ruby-3.0.0.0 has great performance and upgrade advantages.

I look forward to hearing from you.

Best regards

File: Dockerfile (hunk 3 above)
Comment (JuanRodenas) [reply Dockerfile:57]: The implementation of HEALTHCHECK, I have implemented it for the possibility of scalability and for health states of the same.

File: Dockerfile (hunk 2 above)
Comment (JuanRodenas) [reply Dockerfile:60]: The main purpose of CMD is to provide defaults for an executable run, in which case it specifies in the ENTRYPOINT. If desired we can remove it.

**Comment by dometto** [comment]
@JuanRodenas there are still changes to the `Dockerfile` that I requested to be made, like removing redis. Could you please update this PR so that it only changes the user that gollum is run under, and updates to ruby v.30?

**Comment by dometto** [comment]
@JuanRodenas I needed the update of the ruby image for another PR to work, so wanted to get this merged ASAP. I thought I'd just take the necessary changes to the Dockerfile and commit them [in this new PR](https://github.com/gollum/gollum/pull/1889), but could not find a way to credit the commit to you since I think your email is invisible to the public. Since I wanted you to be credited for the commit  I decided to remove all the changes from this PR that we don't want/need, as I saw you enabled the "[allow changes to a PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)" functionality. However, I just noticed that this PR was not based on a PR/feature branch in your repo, but on your `master`!  I'm sorry if this causes problems for you. Of course feel free revert my commits on your repo.

**Comment by dometto** [comment]
@JuanRodenas I suggest that you submit a new PR based off a feature branch on your repo, containing only the changes to the Dockerfile that I've left in this PR.

**Comment by dometto** [comment]
Managed to properly credit @JuanRodenas in #1894 so closing this.
